api/processors/services/group_builder.py

```diff
@@ -146,23 +150,120 @@ def _get_allowed_group_size(age_range: AgeRange) -> GroupSizeRestriction:
+        teacher: Teacher,
+        age_range: AgeRange,
+        language_and_level: LanguageAndLevel,
+        day_time_slots: Collection[DayAndTimeSlot],
+    ) -> GroupCandidate | None:
+        restrictions = GroupBuilder._get_allowed_group_size(age_range)
+        communication_language = CommunicationLanguageMode(
+            teacher.personal_info.communication_language_mode
+        )
+        students = GroupBuilder._get_students(
+            language_and_level=language_and_level,
+            age_range=age_range,
+            communication_language=communication_language,
+            day_time_slots=day_time_slots,
+            max_students_num=restrictions.max,
+        )
+        logger.debug(students)
+        if len(students) < restrictions.min:
+            return None
 
         return GroupCandidate(
-            age_range=AgeRange(age_from=18, age_to=90),
-            language_and_level=teacher.teaching_languages_and_levels.first(),  # type: ignore
-            communication_language_mode=CommunicationLanguageMode(
-                teacher.personal_info.communication_language_mode
-            ),
+            age_range=age_range,
+            language_and_level=language_and_level,
+            communication_language_mode=communication_language,
             teacher=teacher,
-            students=list(range(1, 10)),  # type: ignore
-            monday=timezone.now(),
-            wednesday=timezone.now(),
+            students=students,
+            day_time_slots=day_time_slots,
+        )
+
+    @staticmethod
+    def _get_students(
+        language_and_level: LanguageAndLevel,
+        age_range: AgeRange,
+        communication_language: CommunicationLanguageMode,
+        day_time_slots: Iterable[DayAndTimeSlot],
+        max_students_num: int,
+    ) -> Collection[Student]:
+        # TODO: handle age ranges properly (add more ranges when necessary)
+        student_manager = Student.objects.filter(
+            project_status=StudentProjectStatus.NO_GROUP_YET,
+            teaching_languages_and_levels=language_and_level,
+            personal_info__communication_language_mode=communication_language,
+            age_range__age_from__gte=age_range.age_from,
+            age_range__age_to__lte=age_range.age_to,
         )
+        for time_slot in day_time_slots:
+            student_manager = student_manager.filter(availability_slots=time_slot)
+
+        return student_manager.order_by("personal_info__date_and_time_added")[:max_students_num]
+
+    @staticmethod
+    def _iterate_lesson_times(teacher: Teacher) -> Iterator[tuple[DayAndTimeSlot, DayAndTimeSlot]]:
+        """
+        Iterates tuples of available time slots.
+        There should be at least one free day between lessons.
+        """
+        teacher_availability_slots = teacher.availability_slots.all()
+        for first_availability in teacher_availability_slots:
+            for second_availability in teacher_availability_slots:
```

Comment:
Good place to use itertools.product:

```python
for first_availability, second_availabilty in itertools.product(teacher_availability_slots):
    ....
```

Reply:
I argue if that's more readable but I'll give you that. I know there's rarely a moment when one can legitimately use itertools!

Reply:
> I argue if that's more readable but I'll give you that. I know there's rarely a moment when one can legitimately use itertools!

[I'll give you that](https://www.youtube.com/watch?v=R7GeKLE0x3s&t=665s) 😆